File: domains/video/text_overlay.py
"""Tool ``add_text_overlay``: escreve um texto (título, descrição) sobre o vídeo."""

from __future__ import annotations

from typing import TYPE_CHECKING, Literal, TypedDict

from core.errors import ErrorPayload, ToolError, guarded
from core.fonts import escape_drawtext, escape_filter_path, require_font
from core.jobs import JobSubmitted

if TYPE_CHECKING:
    from mcp.server.mcpserver import MCPServer

    from domains import Runtime

type TextPosition = Literal["top", "center", "bottom"]

_POSITIONS: dict[TextPosition, str] = {
    "top": "h*0.08",
    "center": "(h-text_h)/2",
    "bottom": "h-text_h-h*0.08",
}
_MAX_TEXT = 500


class AddTextOverlayResult(TypedDict):
    """Vídeo gerado com o texto."""

    output: str
    text: str
    position: TextPosition
    start: float
    end: float | None


def _do_overlay(
    runtime: Runtime,
    path: str,
    text: str,
    *,
    position: TextPosition,
    start: float,
    end: float | None,
    font_size: int,
    color: str,
) -> AddTextOverlayResult:
    source = runtime.workspace.existing(path)
    text = text.strip()
    if not text or len(text) > _MAX_TEXT:
        raise ToolError(
            f"text deve ter entre 1 e {_MAX_TEXT} caracteres.",
            code="invalid_argument",
            hint="Quebre textos longos em linhas com \\n ou use burn_subtitles.",
        )
    if font_size <= 0:
        raise ToolError("font_size deve ser maior que zero.", code="invalid_argument")
    if start < 0 or (end is not None and end <= start):
        raise ToolError(
            f"Intervalo inválido: start={start}, end={end}.",
            code="invalid_argument",
            hint="start deve ser >= 0 e end maior que start, ou omitido para ir até o fim.",
        )
    info = runtime.ffmpeg.probe(source)
    if start >= info["duration"]:
        raise ToolError(
            f"start={start}s ultrapassa a duração do vídeo ({info['duration']:.2f}s).",
            code="invalid_argument",
            hint="Use probe_video para conferir a duração.",
        )
    font = require_font()
    enable = f"between(t\\,{start}\\,{end})" if end is not None else f"gte(t\\,{start})"
    drawtext = (
        f"drawtext=fontfile='{escape_filter_path(font)}':text='{escape_drawtext(text)}'"
        f":fontsize={font_size}:fontcolor={color}:borderw=3:bordercolor=black@0.8"
        f":x=(w-text_w)/2:y={_POSITIONS[position]}:line_spacing=8:enable='{enable}'"
    )
    output = runtime.workspace.output_for(source, "text")
    audio_args = ["-c:a", "copy"] if info["has_audio"] else []
    runtime.ffmpeg.run(
        [
            "-i",
            str(source),
            "-vf",
            drawtext,
            "-c:v",
            "libx264",
            "-pix_fmt",
            "yuv420p",
            *audio_args,
            str(output),
        ]
    )
    return AddTextOverlayResult(
        output=runtime.workspace.relative(output),
        text=text,
        position=position,
        start=start,
        end=end,
    )


@guarded
def add_text_overlay(
    runtime: Runtime,
    path: str,
    text: str,
    *,
    position: TextPosition = "bottom",
    start: float = 0.0,
    end: float | None = None,
    font_size: int = 42,
    color: str = "white",
    background: bool = False,
) -> AddTextOverlayResult | JobSubmitted:
    """Implementação pura, testável sem MCP."""
    if background:
        return runtime.jobs.submit(
            "add_text_overlay",
            lambda: _do_overlay(
                runtime,
                path,
                text,
                position=position,
                start=start,
                end=end,
                font_size=font_size,
                color=color,
            ),
        )
    return _do_overlay(
        runtime,
        path,
        text,
        position=position,
        start=start,
        end=end,
        font_size=font_size,
        color=color,
    )


def register(mcp: MCPServer[None], runtime: Runtime) -> None:
    """Expõe a tool no servidor."""

    @mcp.tool(name="add_text_overlay")
    def _tool(  # noqa: PLR0917  # a assinatura da tool é a interface do agente
        path: str,
        text: str,
        position: TextPosition = "bottom",
        start: float = 0.0,
        end: float | None = None,
        font_size: int = 42,
        color: str = "white",
        background: bool = False,
    ) -> AddTextOverlayResult | JobSubmitted | ErrorPayload:
        r"""Escreve um texto fixo sobre o vídeo (título, descrição, chamada) e salva novo arquivo.

        Serve para colocar a descrição ou o título que você escreveu direto na imagem.
        O texto fica centralizado horizontalmente, com contorno preto para leitura.
        Use \\n para quebrar linhas. Para falas sincronizadas use burn_subtitles.
        O original não é modificado; o vídeo é re-encodado.

        Args:
            path: Vídeo, relativo ao workspace.
            text: Texto a exibir. Até 500 caracteres.
            position: top, center ou bottom.
            start: Segundo em que o texto aparece.
            end: Segundo em que o texto some. Omitido = até o fim.
            font_size: Tamanho da fonte em pixels.
            color: Cor do texto (white, yellow, #FF0000...).
            background: Executa como job e devolve job_id.
        """
        return add_text_overlay(
            runtime,
            path,
            text,
            position=position,
            start=start,
            end=end,
            font_size=font_size,
            color=color,
            background=background,
        )

File: version-in-rust/domains/video/text_overlay.rs
//! Tool `add_text_overlay`: escreve um texto (título, descrição) sobre o vídeo.

use std::sync::Arc;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::errors::{guarded, ErrorCode, ToolError, ToolResult};
use crate::core::fonts::{escape_drawtext, escape_filter_path, require_font};
use crate::core::jobs::MaybeJob;
use crate::domains::{McpServer, Runtime};
use crate::ffargs;

const MAX_TEXT: usize = 500;

/// Posição vertical do texto.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "lowercase")]
pub enum TextPosition {
    Top,
    Center,
    Bottom,
}

impl TextPosition {
    /// Nome da posição.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Top => "top",
            Self::Center => "center",
            Self::Bottom => "bottom",
        }
    }

    fn y_expr(self) -> &'static str {
        match self {
            Self::Top => "h*0.08",
            Self::Center => "(h-text_h)/2",
            Self::Bottom => "h-text_h-h*0.08",
        }
    }
}

/// Vídeo gerado com o texto.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct AddTextOverlayResult {
    pub output: String,
    pub text: String,
    pub position: TextPosition,
    pub start: f64,
    pub end: Option<f64>,
}

/// Opções de estilo e tempo do texto.
#[derive(Debug, Clone, PartialEq)]
pub struct OverlayOptions {
    pub position: TextPosition,
    pub start: f64,
    pub end: Option<f64>,
    pub font_size: i64,
    pub color: String,
}

impl Default for OverlayOptions {
    fn default() -> Self {
        Self {
            position: TextPosition::Bottom,
            start: 0.0,
            end: None,
            font_size: 42,
            color: "white".to_string(),
        }
    }
}

fn do_overlay(
    runtime: &Runtime,
    path: &str,
    text: &str,
    options: &OverlayOptions,
) -> ToolResult<AddTextOverlayResult> {
    let source = runtime.workspace.existing(path)?;
    let text = text.trim();
    if text.is_empty() || text.chars().count() > MAX_TEXT {
        return Err(ToolError::with_hint(
            format!("text deve ter entre 1 e {MAX_TEXT} caracteres."),
            ErrorCode::InvalidArgument,
            "Quebre textos longos em linhas com \\n ou use burn_subtitles.",
        ));
    }
    if options.font_size <= 0 {
        return Err(ToolError::new(
            "font_size deve ser maior que zero.",
            ErrorCode::InvalidArgument,
        ));
    }
    let (start, end) = (options.start, options.end);
    if start < 0.0 || end.is_some_and(|e| e <= start) {
        return Err(ToolError::with_hint(
            format!(
                "Intervalo inválido: start={start}, end={}.",
                end.map_or("None".to_string(), |e| e.to_string())
            ),
            ErrorCode::InvalidArgument,
            "start deve ser >= 0 e end maior que start, ou omitido para ir até o fim.",
        ));
    }
    let info = runtime.ffmpeg.probe(&source)?;
    if start >= info.duration {
        return Err(ToolError::with_hint(
            format!(
                "start={start}s ultrapassa a duração do vídeo ({:.2}s).",
                info.duration
            ),
            ErrorCode::InvalidArgument,
            "Use probe_video para conferir a duração.",
        ));
    }
    let font = require_font()?;
    let enable = match end {
        Some(end) => format!("between(t\\,{start}\\,{end})"),
        None => format!("gte(t\\,{start})"),
    };
    let drawtext = format!(
        "drawtext=fontfile='{}':text={}:fontsize={}:fontcolor={}:borderw=3:bordercolor=black@0.8\
         :x=(w-text_w)/2:y={}:line_spacing=8:enable='{enable}'",
        escape_filter_path(&font),
        escape_drawtext(text),
        options.font_size,
        options.color,
        options.position.y_expr()
    );
    let output = runtime.workspace.output_for(&source, "text", None);
    let mut args = ffargs!["-i", source, "-vf", drawtext, "-c:v", "libx264", "-pix_fmt", "yuv420p"];
    if info.has_audio {
        args.extend(ffargs!["-c:a", "copy"]);
    }
    args.push(output.clone().into_os_string());
    runtime.ffmpeg.run(&args)?;
    Ok(AddTextOverlayResult {
        output: runtime.workspace.relative(&output),
        text: text.to_string(),
        position: options.position,
        start,
        end,
    })
}

/// Implementação pura, testável sem MCP.
///
/// # Errors
///
/// Texto vazio ou longo demais, intervalo inválido, arquivo inexistente ou falha do ffmpeg.
pub fn add_text_overlay(
    runtime: &Arc<Runtime>,
    path: &str,
    text: &str,
    options: OverlayOptions,
    background: bool,
) -> ToolResult<MaybeJob<AddTextOverlayResult>> {
    if background {
        let runtime_job = Arc::clone(runtime);
        let path = path.to_string();
        let text = text.to_string();
        return Ok(MaybeJob::Job(
            runtime.jobs.submit("add_text_overlay", move || {
                do_overlay(&runtime_job, &path, &text, &options)
            }),
        ));
    }
    Ok(MaybeJob::Done(do_overlay(runtime, path, text, &options)?))
}

/// Parâmetros da tool.
#[derive(Debug, Deserialize, JsonSchema)]
pub struct Params {
    /// Vídeo, relativo ao workspace.
    pub path: String,
    /// Texto a exibir. Até 500 caracteres.
    pub text: String,
    /// top, center ou bottom.
    #[serde(default = "default_position")]
    pub position: TextPosition,
    /// Segundo em que o texto aparece.
    #[serde(default)]
    pub start: f64,
    /// Segundo em que o texto some. Omitido = até o fim.
    #[serde(default)]
    pub end: Option<f64>,
    /// Tamanho da fonte em pixels.
    #[serde(default = "default_font_size")]
    pub font_size: i64,
    /// Cor do texto (white, yellow, #FF0000...).
    #[serde(default = "default_color")]
    pub color: String,
    /// Executa como job e devolve job_id.
    #[serde(default)]
    pub background: bool,
}

fn default_position() -> TextPosition {
    TextPosition::Bottom
}

fn default_font_size() -> i64 {
    42
}

fn default_color() -> String {
    "white".to_string()
}

/// Expõe a tool no servidor.
pub fn register(mcp: &mut McpServer, runtime: &Arc<Runtime>) {
    let runtime = Arc::clone(runtime);
    mcp.tool(
        "add_text_overlay",
        "Escreve um texto fixo sobre o vídeo (título, descrição, chamada) e salva novo arquivo.\n\n\
         Serve para colocar a descrição ou o título que você escreveu direto na imagem. \
         O texto fica centralizado horizontalmente, com contorno preto para leitura. \
         Use \\n para quebrar linhas. Para falas sincronizadas use burn_subtitles. \
         O original não é modificado; o vídeo é re-encodado.",
        move |params: Params| {
            guarded(add_text_overlay(
                &runtime,
                &params.path,
                &params.text,
                OverlayOptions {
                    position: params.position,
                    start: params.start,
                    end: params.end,
                    font_size: params.font_size,
                    color: params.color,
                },
                params.background,
            ))
        },
    );
}
